packages/react-reconciler/src/__tests__/ReactSuspenseList-test.internal.js

```diff
@@ -1089,4 +1088,503 @@ describe('ReactSuspenseList', () => {
+        </SuspenseList>
+      );
+    }
+
+    ReactNoop.render(<Foo />);
+
+    expect(() => Scheduler.unstable_flushAll()).toWarnDev([
+      'Warning: "collapse" is not a supported value for tail on ' +
+        '<SuspenseList />. Did you mean "collapsed"?' +
+        '\n    in SuspenseList (at **)' +
+        '\n    in Foo (at **)',
+    ]);
+  });
+
+  it('warns if a tail option is used with "together"', () => {
+    function Foo() {
+      return (
+        <SuspenseList revealOrder="together" tail="collapsed">
+          <Suspense fallback="Loading">Content</Suspense>
+        </SuspenseList>
+      );
+    }
+
+    ReactNoop.render(<Foo />);
+
+    expect(() => Scheduler.unstable_flushAll()).toWarnDev([
+      'Warning: <SuspenseList tail="collapsed" /> is only valid if ' +
+        'revealOrder is "forwards" or "backwards". ' +
+        'Did you mean to specify revealOrder="forwards"?' +
+        '\n    in SuspenseList (at **)' +
+        '\n    in Foo (at **)',
+    ]);
+  });
+
+  it('adding to the middle does not collapse insertions (forwards)', async () => {
+    let A = createAsyncText('A');
+    let B = createAsyncText('B');
+    let C = createAsyncText('C');
+    let D = createAsyncText('D');
+    let E = createAsyncText('E');
+    let F = createAsyncText('F');
+
+    function Foo({items}) {
+      return (
+        <SuspenseList revealOrder="forwards" tail="collapsed">
+          {items.map(([key, Component]) => (
+            <Suspense key={key} fallback={<Text text={'Loading ' + key} />}>
+              <Component />
+            </Suspense>
+          ))}
+        </SuspenseList>
+      );
+    }
+
+    ReactNoop.render(<Foo items={[['A', A], ['D', D]]} />);
+
+    await A.resolve();
+    await D.resolve();
+
+    expect(Scheduler).toFlushAndYield(['A', 'D']);
+
+    // First render commits A and D.
+    expect(ReactNoop).toMatchRenderedOutput(
+      <Fragment>
+        <span>A</span>
+        <span>D</span>
+      </Fragment>,
+    );
+
+    // For the second render, we're going to insert items in the middle and end.
+    ReactNoop.render(
+      <Foo
+        items={[['A', A], ['B', B], ['C', C], ['D', D], ['E', E], ['F', F]]}
+      />,
+    );
+
+    expect(Scheduler).toFlushAndYield([
+      'A',
+      'Suspend! [B]',
+      'Loading B',
+      'Suspend! [C]',
+      'Loading C',
+      'D',
+      'Loading E',
+    ]);
+
+    // B and C don't get collapsed, but F gets collapsed with E.
+    expect(ReactNoop).toMatchRenderedOutput(
+      <Fragment>
+        <span>A</span>
+        <span>Loading B</span>
+        <span>Loading C</span>
+        <span>D</span>
+        <span>Loading E</span>
+      </Fragment>,
+    );
+
+    await B.resolve();
+
+    expect(Scheduler).toFlushAndYield(['B', 'Suspend! [C]']);
+
+    // Incremental loading is suspended.
+    jest.advanceTimersByTime(500);
+
+    // Even though B is unsuspended, it's still in loading state because
+    // it is blocked by C.
```

Comment:
Seems counter-intuitive to me that B is blocked by C in this case. Since we display both loading states, I would also have expected them to resolve in order.

Reply:
This isn't new behavior in this PR and this is basically just replicating the same assertion as other tests. The rationale for this behavior is that in a variable size list you don't want to pop in things multiple time to shift the below content around multiple times. Therefore the "head" resolves in the "together" mode.

This is kind of a rare case though since insertions in the middle are rare. Let alone multiple of them. A more common variant is inserting at the top but in that case you don't want to start with the top most and then expand them one by one to shift things down. More likely it's better to just show them all as a unit.